core/src/replay_stage.rs

```diff
@@ -2511,6 +2516,8 @@ impl ReplayStage {
                 })
                 .unwrap_or_else(|err| warn!("Error: {:?}", err));
             last_vote_refresh_time.last_refresh_time = Instant::now();
+        } else {
+            tower.last_vote_tx_blockhash_failed();
```

Comment:
hmm we only want to mark this as failed if this is a non-voter right? It seems like there's a few other cases where `generate_vote_tx` can return `None` though. 

Same is true for the other place where this is called

Reply:
non-voter is just one example. `generate_vote_tx` could fail for multiple reasons. In that case the vote is already in the tower under `last_vote` but we haven't propagated it to the network.

I think there's no point attempting to refresh the vote if the last `generate_vote_tx` failed as the `generate_vote_tx` for the refreshed vote should fail again. In steady state all of the checks in that function should be deterministic.

However if you prefer I don't mind keeping the scope to just the `authorized_voter_keypair` check, which is what the `--no-voting` flag does.

Reply:
hmmm, seems like we should differentiate a failed vote from somebody that's never voting. Grouping them together seems bound to cause avoid future confusion/bugs.

I would limit the scope if possible